### ISSUE TITLE
Custom iterators should implement `#size`

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#11019]
If a custom iterator does not implement `#size`, calling that methods falls back to `Enumerable#size` which is just `count { true }`. That requires iterating the entire collection and thus completely busts the purpose of an iterator.

Example: `(1..60).to_a.each_combination(53).size` takes very long to execute because all combinations need to be iterated, even though the size could be calculated instantaneously. See https://forum.crystal-lang.org/t/array-combinations-issues/3557/15

Contributing factors are:
* https://github.com/crystal-lang/crystal/issues/10014
* https://github.com/crystal-lang/crystal/issues/9998#issuecomment-737629432

If either of those issues was resolved, the compiler would force every `Iterator` to implement `#size`.

Even without pushing from the compiler, we should make sure to add `#size` methods to all `Iterator` implementations.

For iterators that directly map their items to an original collection (`Indexable::ItemIterator`, `Int::TimesIterator`), this is a trivial forward. Combinatory iterators like (`Indexable::CombinationIterator`) can calculate their size from the original collection in constant time.

Some iterators don't know their size in advance (`Dir::EntryIterator`, `Iterator::Select`) or are infinite (`Iterator::Singleton`). There we're at the heart of the discussion in #10014.

**Comment by HertzDevil** [comment]
What do you think the following should output?

```crystal
iter = (1..60).to_a.each_combination(53)
iter.next
iter.size
```

**Comment by asterite** [comment]
https://stackoverflow.com/a/1838732

**Comment by straight-shoota** [comment]
In Ruby it's the total size.

```ruby
(1..60).to_a.combination(53).tap(&:next).size # => 386206920
```

And I think that makes sense. `#size` returns *the number of elements in the collection.* It's not the number of elements left to be iterated.

**Comment by asterite** [comment]
Right, that's what I also proposed some time ago, that iterators to_a, map, size, etc., start from the beginning. I think now it's too late.

**Comment by kostya** [comment]
Why not add new method `estimated_size`, which would always return full iterator size. For all it would implement as equal size, but some iterators reimplement it.

**Comment by asterite** [comment]
https://github.com/crystal-lang/crystal/pull/10089

**Comment by HertzDevil** [comment]
> Why not add new method `estimated_size`, which would always return full iterator size. For all it would implement as equal size, but some iterators reimplement it. I think you need this to preallocate.

Related: #10075

The argument there is that just because some returned container is preallocated with the iterator's estimated size doesn't mean the preallocation always gives optimal performance, because it changes the reallocation behaviour of subsequent inserts.
